File: src/main/java/org/mech/rougue/core/r/model/map/Map.java
package org.mech.rougue.core.r.model.map;

import java.util.ArrayList;
import java.util.List;

import org.mech.rogue.game.export.Exportable;
import org.mech.rogue.game.model.map.MapTile;
import org.mech.rogue.game.model.map.TileConfig;
import org.mech.rougue.core.game.model.map.MapStats;
import org.mech.rougue.core.game.model.map.tile.Tiles;
import org.mech.rougue.core.r.model.area.Area;
import org.mech.rougue.core.r.model.common.GObject;
import org.mech.rougue.utils.CollectionUtils;
import org.mech.terminator.geometry.Dimension;
import org.mech.terminator.geometry.Position;

public class Map implements Exportable {
	private static final long serialVersionUID = 6589737997245185262L;

	private List<GObject> gameObjects;
	private String mapId;
	private MapTile[][] newArea;
	private Dimension mapSize;
	private MapStats stats;
	private Area area;

	public Map(final Dimension mapSize) {
		this(mapSize, Tiles.VOID);
	}

	public Map(final Dimension mapSize, final TileConfig defaultTileGround) {
		this.mapSize = mapSize;
		this.stats = new MapStats(this);
		this.newArea = new MapTile[mapSize.width][mapSize.height];

		for (int i = 0; i < mapSize.width; i++) {
			for (int j = 0; j < mapSize.height; j++) {
				//TODO renderId
				final MapTile newMapTile = new MapTile(defaultTileGround);
				newArea[i][j] = newMapTile;
			}
		}

		gameObjects = new ArrayList<GObject>();
	}

	public MapTile get(final int x, final int y) {
		if (x < 0 || y < 0 || x >= mapSize.width || y >= mapSize.height) {
			return null;
		}
		return newArea[x][y];
	}

	public MapTile get(final Position p) {
		return get(p.x, p.y);
	}

	public void put(final MapTile tile, final int x, final int y) {
		newArea[x][y] = tile;
	}

	public void put(final MapTile tile, final Position p) {
		put(tile, p.x, p.y);
	}

	public Dimension getSize() {
		return mapSize;
	}

	public void add(final GObject gameObject) {
		this.gameObjects.add(gameObject);
	}

	public void remove(final GObject gameObject) {
		this.gameObjects.remove(gameObject);
	}

	public void addAll(final List<GObject> objects) {
		if (CollectionUtils.isNotEmpty(objects)) {
			for (final GObject object : objects) {
				add(object);
			}
		}
	}

	public MapStats getStats() {
		return stats;
	}

	public boolean isPositionInMap(final Position position) {
		return !(position.x < 0 || position.x >= newArea.length || position.y < 0 || position.y >= newArea[0].length);
	}

	public void setStats(final MapStats mapStats) {
		this.stats = mapStats;
	}

	public void handleEnvironmentChange(final EnvironmentObject eObject) {
		final MapTile tile = get(eObject.getPosition());
		// TODO
		// tile.setPassable(eObject.isPassable());
		// tile.setObstacle(eObject.isLightObstacle());
	}

	public String getMapId() {
		return mapId;
	}

	public void setMapId(final String mapId) {
		this.mapId = mapId;
	}

	public Area getArea() {
		return area;
	}

	public void setArea(final Area area) {
		this.area = area;
	}

	public List<GObject> getGameObjects() {
		return gameObjects;
	}

	@Override
	public String objectId() {
		return mapId;
	}
}

File: src/main/scala/org/mech/rogue/game/model/map/Map.scala
package org.mech.rogue.game.model.map

import org.mech.rogue.game.export.Exportable
import org.mech.rogue.game.render.map.RenderObject
import org.mech.rougue.core.r.`object`.GObjectUtils
import org.mech.rougue.core.r.model.area.Area
import org.mech.rougue.core.r.model.common.GObject
import org.mech.terminator.geometry.{Dimension, Position}

import scala.collection.mutable.ListBuffer

class Map(val mapId: String, val size: Dimension, val area: Area, defaultTile: TileConfig) extends Exportable {
  private val gameObjects: ListBuffer[GObject] = new ListBuffer[GObject]
  val tiles: Array[Array[MapTile]] = Array.fill(size.width, size.height)(new MapTile(defaultTile))
  val stats = new Stats(this)

  override def objectId: String = mapId

  def add(gameObject: GObject): Unit = gameObjects += gameObject

  def remove(gameObject: GObject): Unit = gameObjects -= gameObject

  def get(pos: Position): Option[MapTile] = {
    if (size.isIn(pos)) {
      Some(tiles(pos.x)(pos.y))
    } else {
      None
    }
  }

  def objects(): List[GObject] = gameObjects.toList

  def get(x: Int, y: Int): Option[MapTile] = get(Position.at(x, y))

  def put(tile: MapTile, pos: Position): Unit = put(tile, pos.x, pos.y)

  def put(tile: MapTile, x: Int, y: Int): Unit = tiles(x)(y) = tile
}

object Map {
  def getObjects[T <: RenderObject](map: Map, pos: Position, clazz: Class[T]): List[T] = {
    // TODO map maybe should store objects better way
    import scala.collection.JavaConverters._
    GObjectUtils.getObjectsOfType(map.objects().asJava, clazz)
      .asScala.filter(obj => obj.getPosition == pos).toList
  }


}
